### ISSUE TITLE
Move to 4.0.3 breaks ABI compatability ?

### ISSUE THREAD

**Issue by amckinstry** [open-mpi/ompi#7616]
This is an issue i'm seeing on Debian unstable, which has recently moved to 4.0.3

Our regression tests show a failure with mpigrafic:
```
mpgrafic: symbol lookup error: mpgrafic: undefined symbol: ompi_f08_mpi_double_precision
```
This is with a move from 4.0.2 -> 4.0.3:
https://ci.debian.net/data/autopkgtest/testing/amd64/m/mpgrafic/4897659/log.gz

This appears to be because symbol(s) were removed:

https://github.com/open-mpi/ompi/commit/51e23f8cb6c7e9a663462923e55eafc3180dbfd5#diff-8c7fd97572800eac0b55c84f29ee9451

I think the unused symbols were not as unused as you expected.

**Comment by gpaulsen** [comment]
Added blocker label to block a v4.0.x release until we understand this, and hopefully fix this.

**Comment by gpaulsen** [comment]
@jsquyres said he'd take a look at this.  We probably just want a solution on the v4.0.x branch, and leave master as-is for upcoming v5.0.

FYI:  v4.0.x PR looks like this came into v4.0.x on https://github.com/open-mpi/ompi/pull/7116/

**Comment by ggouaillardet** [comment]
@amckinstry I am now double checking in my environment.

FWIW, the commit you are pointing is only for the `master` branch.
In order to workaround a `flang` bug/feature, some constants were revamped (e.g. moved from a `BIND(C, ...` symbol to a `parameter`), so that clearly broke the ABI and we are aware of it.
That being said, extra care was taken when back porting this to the `v4.0.x` branch.
Newly compiled programs will use the `parameter`, but the previous symbols were kept not to break the ABI.

What if you
```
nm libmpi_usempif08.so | grep double_precision
```
could the symbol be there but not visible?
(Debian's linker is way stricter than the one in RedHat I mainly use)

**Comment by ggouaillardet** [comment]
the root cause is ompi/mpi/fortran/use-mpi-f08/base/constants.c has been removed from Makefile.am, so the symbols are no more slurped into the libmpi_usempif08.so library.

It seems that was caused by subsequent backport(s), I am now working on a fix.

**Comment by ggouaillardet** [comment]
@amckinstry you can manually download and apply the patch at https://github.com/open-mpi/ompi/pull/7622.patch.
If you apply it on the `v4.0.3` tarball, you will also need to run `./autogen.pl --force`

**Comment by jsquyres** [comment]
Thanks @ggouaillardet

**Comment by jsquyres** [comment]
@amckinstry The fix for this was merged in #7622.  Can you try the latest nightly v4.0.x snapshot to make sure we got it right?

https://www.open-mpi.org/nightly/v4.0.x/

**Comment by gpaulsen** [comment]
@amckinstry, Thanks.  Based on the sounds of it, this should possibly drive an earlier v4.0.4 release, would you agree?

**Comment by amckinstry** [comment]
The patch appears to be missing constants.c (not present in 4.0.3) ? Adding constants.[ch] makes the patch work.
Meanwhile my test case has disappeared, as mpgrafic in its latest version (0.3.18) had changed to avoid using the symbols. No urgency on at 4.0.4 ?

**Comment by ggouaillardet** [comment]
@amckinstry sorry for the glitch.

Indeed, since `constants.c` was missing from `Makefile.am`, it has not been included into the tarball.
`constants.h` is automatically generated when running `autogen.pl` (`autogen.pl --force` in your case since you are running from a tarball)

**Comment by jsquyres** [comment]
As far as I can tell, the issue has been resolved on the v4.0.x branch.  I'm closing this issue -- let us know if we need to re-open it.
